File: examples/claude_code_hook.py
#!/usr/bin/env python3
"""Sentinel AI hook for Claude Code.

Use as a pre-tool-call hook in Claude Code to scan tool arguments
for safety issues before execution.

Setup:
    Add to your Claude Code settings (~/.claude/settings.json):

    {
      "hooks": {
        "PreToolUse": [
          {
            "matcher": "Bash",
            "command": "python /path/to/claude_code_hook.py"
          }
        ]
      }
    }

    The hook reads the tool call from stdin (JSON) and outputs
    a decision to stdout. If unsafe content is detected, it
    blocks the tool call with a reason.

Protocol:
    Input (stdin):  JSON with tool_name and tool_input fields
    Output (stdout): JSON with decision ("allow", "block", or "review")
                     and optional reason
"""

import json
import sys

from sentinel.scanners.tool_use import ToolUseScanner
from sentinel.scanners.pii import PIIScanner
from sentinel.core import RiskLevel


def main():
    # Read tool call from stdin
    try:
        raw = sys.stdin.read()
        tool_call = json.loads(raw)
    except (json.JSONDecodeError, KeyError):
        # Can't parse input — allow by default
        print(json.dumps({"decision": "allow"}))
        return

    tool_name = tool_call.get("tool_name", "")
    tool_input = tool_call.get("tool_input", {})

    # Initialize scanners
    tool_scanner = ToolUseScanner()
    pii_scanner = PIIScanner()

    # Scan the tool call
    findings = tool_scanner.scan_tool_call(tool_name, tool_input)

    # Also check for PII in arguments
    text_parts = []
    _collect_strings(tool_input, text_parts)
    combined_text = " ".join(text_parts)
    pii_findings = pii_scanner.scan(combined_text)

    all_findings = findings + pii_findings

    if not all_findings:
        print(json.dumps({"decision": "allow"}))
        return

    max_risk = max(f.risk for f in all_findings)

    if max_risk >= RiskLevel.HIGH:
        reasons = [f"{f.description} ({f.risk.value})" for f in all_findings]
        print(json.dumps({
            "decision": "block",
            "reason": f"Sentinel AI blocked this tool call: {'; '.join(reasons)}",
        }))
    elif max_risk >= RiskLevel.MEDIUM:
        reasons = [f"{f.description} ({f.risk.value})" for f in all_findings]
        print(json.dumps({
            "decision": "review",
            "reason": f"Sentinel AI flagged for review: {'; '.join(reasons)}",
        }))
    else:
        print(json.dumps({"decision": "allow"}))


def _collect_strings(obj, parts):
    if isinstance(obj, str):
        parts.append(obj)
    elif isinstance(obj, dict):
        for v in obj.values():
            _collect_strings(v, parts)
    elif isinstance(obj, (list, tuple)):
        for item in obj:
            _collect_strings(item, parts)


if __name__ == "__main__":
    main()

File: examples/claude_code_hook.ts
/**
 * Sentinel AI — Claude Code PostToolUse Hook (TypeScript)
 *
 * Scans every Write/Edit tool output for:
 * - OWASP vulnerabilities (SQL injection, XSS, command injection, etc.)
 * - Supply chain attacks in package manifests
 * - Hardcoded secrets and API keys
 *
 * Setup — add to .claude/settings.json:
 * {
 *   "hooks": {
 *     "PostToolUse": [
 *       {
 *         "matcher": "Write|Edit",
 *         "hooks": [{
 *           "type": "command",
 *           "command": "npx tsx examples/claude_code_hook.ts"
 *         }]
 *       }
 *     ]
 *   }
 * }
 *
 * Run: reads tool output from stdin, exits 2 to block on high/critical findings
 */

import { CodeScanner, DependencyScanner } from '../sdk-js/src/index';

const MANIFEST_FILES = new Set([
  'requirements.txt', 'package.json', 'pyproject.toml',
  'Pipfile', 'setup.py', 'setup.cfg',
]);

async function main(): Promise<number> {
  // Read stdin
  let raw = '';
  try {
    const chunks: Buffer[] = [];
    for await (const chunk of process.stdin) {
      chunks.push(chunk);
    }
    raw = Buffer.concat(chunks).toString('utf-8');
  } catch {
    return 0;
  }

  if (!raw.trim()) return 0;

  let event: Record<string, unknown>;
  try {
    event = JSON.parse(raw);
  } catch {
    return 0;
  }

  const toolName = (event.tool_name as string) ?? '';
  const toolInput = (event.tool_input as Record<string, unknown>) ?? {};

  if (!['Write', 'Edit', 'write', 'edit'].includes(toolName)) {
    return 0;
  }

  const filePath = (toolInput.file_path ?? toolInput.path ?? '') as string;
  const content = (toolInput.content ?? toolInput.new_string ?? '') as string;

  if (!content) return 0;

  const allFindings: Array<{ risk: string; description: string }> = [];

  // Code vulnerability scan
  const codeScanner = new CodeScanner();
  allFindings.push(...codeScanner.scan(content, filePath));

  // Dependency scan (if manifest file)
  const fileName = filePath.split('/').pop() ?? filePath.split('\\').pop() ?? '';
  if (MANIFEST_FILES.has(fileName)) {
    const depScanner = new DependencyScanner();
    allFindings.push(...depScanner.scan(content, fileName));
  }

  if (allFindings.length === 0) return 0;

  const criticalHigh = allFindings.filter(f =>
    f.risk === 'CRITICAL' || f.risk === 'HIGH'
  );
  const medium = allFindings.filter(f => f.risk === 'MEDIUM');

  // Warn on medium findings
  for (const f of medium) {
    console.log(`  WARNING: ${f.description}`);
  }

  // Block on high/critical
  if (criticalHigh.length > 0) {
    console.log(`\nSentinel AI: blocked ${toolName} — ${criticalHigh.length} security issue(s) found:\n`);
    for (const f of criticalHigh) {
      console.log(`  [${f.risk}] ${f.description}`);
    }
    console.log(`\n  File: ${filePath}`);
    console.log('  Fix the issues above before writing this code.');
    return 2;
  }

  return 0;
}

main().then(code => process.exit(code));
